Src/DependencyCollector/Shared/Implementation/HttpDiagnosticSourceSubscriber.cs

```diff
@@ -0,0 +1,109 @@
+#if !NET40
+namespace Microsoft.ApplicationInsights.DependencyCollector.Implementation
+{
+    using System;
+    using System.Collections.Concurrent;
+    using System.Collections.Generic;
+    using System.Diagnostics;
+    using System.Globalization;
+    using System.Linq;
+    using System.Net;
+    using Microsoft.ApplicationInsights.Common;
+    using Microsoft.ApplicationInsights.DataContracts;
+    using Microsoft.ApplicationInsights.Extensibility;
+    using Microsoft.ApplicationInsights.Extensibility.Implementation;
+    using Microsoft.ApplicationInsights.Web.Implementation;
+
+    /// <summary>
+    /// A helper subscriber class helping the parent object, which is a HttpDiagnosticSourceListener, to subscribe
+    /// to the Http DiagnosticSource. That way the parent object can subscribe to the DiagnosticSource without worry
+    /// about the details around subscription.
+    /// </summary>
+    internal class HttpDiagnosticSourceSubscriber : IObserver<DiagnosticListener>, IDisposable
+    {
+        private HttpDiagnosticSourceListener parent;
+        private IDisposable allListenersSubscription;
+        private IDisposable sourceSubscription;
+        private bool disposed = false;
+
+        internal HttpDiagnosticSourceSubscriber(HttpDiagnosticSourceListener parent)
+        {
+            GC.KeepAlive(HttpHandlerDiagnosticListener.SingletonInstance);
```

Comment:
Also, is this bit of code going to migrate into Corefx too? If not, then the SingletonInstance presumably needs to be made public.